ocaml/xapi/xapi_xenops.ml

```diff
@@ -2082,8 +2082,14 @@ let update_vm ~__context id =
                   (fun (_, state) ->
                     match state.Vm.domids with
                     | value :: _ ->
-                        Db.VM.set_domid ~__context ~self
-                          ~value:(Int64.of_int value)
+                        if
+                          Db.VM.get_domid ~__context ~self <> Int64.of_int value
```

Comment:
Would the line `( if different (fun x -> x.domids) then` already ensure the difference?

Reply:
No

Reply:
The `different` with cache recognizes the differences already. So here the change is trying to filter out a fake positive? E.g. in cache it is A, but from xenopsd, it is B, and actually it is B ? 
But I understand the case should be considered is E.g. in cache it is A, but from xenopsd, it is B and then A, so that the `different` can't spot the change.

Reply:
And the cache may be empty, e.g. after xapi restarts.

Reply:
In theory, a VM can get the same domid when resumed as it had before it was suspended. For example, if it resumes on another host, or if the host has rebooted. Perhaps we should not rely on the domid change, and just remove the guidance in the `resume` function. You're doing that for migration as well.

Reply:
Yes, a VM might get the same domid when resumed:
1. resumes on another host, I think it is a VM migrate, and will be covered in `pool_migrate` in `xapi_vm_migrate.ml`
2. if the host has rebooted, I think then the VM will be closed first and it's domid will be set to `-1` first, so it will be covered in `force_state_reset_keep_current_operations` in `xapi_vm_lifecycle.ml`

Do you mean to remove the `restart_device_model` guidance in `resume` in `xapi_xenops.ml`? But I think then it will not be triggered when the VM is rebooted by XC or inside itself?

Reply:
Although suspend+resume and live migration use similar mechanisms internally, they are different higher up in xapi. You suspend a VM using the `VM.suspend` call, which saves its memory image to a VDI and writes the xenopsd state to a field in the xapi database. You then resume it using `VM.resume`, on any host in the pool and at any time. No code in xapi_vm_migrate.ml is used here.

But yes, it looks like we are always setting the domid to -1 when entering the suspend state, in `force_state_reset_keep_current_operations`, which would cover both cases. And you are already clearing the guidance fields there. Then the question is: do we really need to catch the changing domid case here again?

Reply:
>Do you mean to remove the restart_device_model guidance in resume in xapi_xenops.ml? But I think then it will not be triggered when the VM is rebooted by XC or inside itself?

We'd still need to watch the `last_start_time` to catch the internal reboot case.

Reply:
[Clearing `restart_device_model` guidance when setting `domid` to -1](https://github.com/gangj/xen-api/blob/private/gangj/CP-45570/ocaml/xapi/xapi_vm_lifecycle.ml#L887-L888) in `force_state_reset_keep_current_operations` covers 2 scenarios: "VM shutdown", "VM suspend". And It does not cover "VM migrate" case. "VM migrate" is covered in `pool_migrate_complete` [here](https://github.com/gangj/xen-api/blob/private/gangj/CP-45570/ocaml/xapi/xapi_vm_migrate.ml#L482-L483).

For "VM shutdown" and "VM suspend", `restart_device_model` will not be cleared again when `domid` change is detected [here](https://github.com/gangj/xen-api/blob/private/gangj/CP-45570/ocaml/xapi/xapi_xenops.ml#L2091-L2092), because `force_state_reset_keep_current_operations` is called [when `power_state` change is detected](https://github.com/gangj/xen-api/blob/private/gangj/CP-45570/ocaml/xapi/xapi_xenops.ml#L2053), which is before `domid` change can be detected, and as `domid` is set to `-1` in `force_state_reset_keep_current_operations` already, so `domid` changing will not be detected for "VM shutdown" and "VM suspend".

If we clear `restart_device_model` guidance when `last_start_time` change is detected(instead of `domid` change is detected) the same as we clear the `restart_vm` guidance now, then for "VM shutdown", the `restart_device_model` guidance will be cleared twice: the first time [when `domid` is set to `-1`](https://github.com/gangj/xen-api/blob/private/gangj/CP-45570/ocaml/xapi/xapi_vm_lifecycle.ml#L887-L888) in `force_state_reset_keep_current_operations`, and then again when `last_start_time` is detected.

Reply:
The overall code change is quite small, but it is hard to get a good overview of the logic. If fields are updated in too many places, then it becomes really hard to understand what is going on. And if we rely too much on "if X happens, then we no longer need to do Y" logic, then I quickly get lost.

It's complicated enough now, while we are working on the feature; imagine looking at this again in two years when fixing some bug. I prefer fewer places where a given field is set, as long as it is overall correct.

I think:

* `force_state_reset` always handles the shutdown and suspend cases
* `pool_migrate_complete` handles live migration
* watching `last_start_time` handles reboots

and

* on shutdown and reboot, all guidance is cleared
* on suspend and live migrate, only restart_device_model is removed

Is tracking domid changes then _necessary_?

Reply:
OK, changed to clear `RestartDeviceModel` guidance on `last_start_time` change, and with this change, clearing `RestartDeviceModel` inside `force_state_reset_keep_current_operations` is not needed: https://github.com/xapi-project/xen-api/pull/5318/commits/09be7631117cad1dd4a35b046b9a33edc21cb0c3